### ISSUE TITLE
Z-Wave SetPoint Unit Conversion Failure

### ISSUE THREAD

**Issue by nayrnet** [domoticz/domoticz#310]
Sometime recently something was broken regarding the SetPoints of Z-Wave thermostats using Fahrenheit

If you set your whole system to C, and input the temp in F, it works.. otherwise the setpoints are entirely broken and any values given are ignored.

It seems somewhere the conversion between units got inverted, I've seen this since I tried an update in mid december.. the changelog shows many changes to SetPoint routines in the time between my updates.

See: http://www.domoticz.com/forum/viewtopic.php?f=6&t=9807
See: http://www.domoticz.com/forum/viewtopic.php?f=6&t=9510
See: http://www.domoticz.com/forum/viewtopic.php?f=31&t=7956

**Comment by gizmocuz** [comment]
Strange, because if i remember correctly the setpoint values had to be send in Celsius to the zwave devices
It looks like we need to send this in Fahrenheit now ?
Are you compiling yourself ? If yes, could you change the function

bool MainWorker::SetSetPointInt

at almost the bottom to:

```
        float tempDest = TempValue;
        if ((pHardware->HwdType != HTYPE_OpenZWave) && (pHardware->HwdType != HTYPE_RazberryZWave))
        {
            unsigned char tSign = m_sql.m_tempsign[0];
            if (tSign == 'F')
            {
                //Convert to Celsius
                tempDest = (tempDest - 32.0f) / 1.8f;
            }
        }

        _tThermostat tmeter;
```

Could you also check from the zwave hardware setup in the control panel (via the node dropdown) if the value displayed there is in Celsius or Fahrenheit ?

**Comment by nayrnet** [comment]
that took care of the issue it appears, perhaps upstream OZW libs now detect what unit the thermostat its self is configured for? dont seem like it.. I used these same OZW libs on an older version of Domo and they worked.

on the backend it looks like its still using C, 22.22C = 72F
Idx Name    Current state   Last updated    sValues
24  Heater Setpoint     2016-01-11 09:48:15 22.22

perhaps there is something deeper inverting the values? Dont know if this is just covering up a bigger underlying issue, will keep testing it.. make sure this dont break it for our metric friends.

**Comment by nayrnet** [comment]
now when I have my system set to C and I feed it C numbers i got the behavior I saw before.. 

If thermostat is set to 22.2, and I change it to 23.. no change on the thermostat
if thermostat is set to 22.2 and I change it to 73c, the thermostat changes from 72F to 73F

so all you did was invert the issue, again.. need to find the source of the issue and fix it there, this is not the solution.

**Comment by gizmocuz** [comment]
okey, please let me know if you find something

**Comment by nayrnet** [comment]
I am looking at the original code of the last working version, and I am not seeing anything about converting to Celsius in the SetSetPoint function.. I suspect this was added later for another thermostat, but curious why the exemption above did not resolve it.

perhaps there is another conversion going on, need to find it... wonder where zw unit conversion happened in this old version, has to be somewhere.

**Comment by nayrnet** [comment]
Found it, ConvertToCelsius & ConvertToFahrenheit in Helper.cpp

But I cant find any calls to it, sigh..

**Comment by gizmocuz** [comment]
Great, i guess this is only for zwave devices as i dont here issues for other devices like toon/nest/thermosmart/icy/....

**Comment by nayrnet** [comment]
Im thinking it was introduced with: 2.3xxx Fixed: Nest Thermostat, setting set point in Fahrenheit

this is my first time really digging into the codebase, going to try to find that commit and see what it all did.

**Comment by nayrnet** [comment]
CNest::SetSetpoint has the exact same routine in it, hrmm.. with the comment: 
        //Maybe this should be done in the main app, so all other devices will also do this

**Comment by gizmocuz** [comment]
Thanks for the help !

**Comment by nayrnet** [comment]
Check this out: https://github.com/domoticz/domoticz/search?utf8=%E2%9C%93&q=tempDest

This dont seem right to me, nest and netamo have there own conversions.. then main does it again. I'll try taking it out of the main thread and see if that fixes it.

**Comment by gizmocuz** [comment]
As none of these thermostats seems to have problems, i suggested the above patch in the mainworker to not convert it for zwave devices

**Comment by nayrnet** [comment]
Well thats just a presumption that it works, alot of the developers are metric and might not have seen it.. or the internal conversion routines for specific devices are hiding the issue.. since I did not find one specific for the zwave hardware

I found a routine called ConvertTemperature, going to try to use it instead

**Comment by nayrnet** [comment]
when I removed that conversion from mainworker.cpp, like my old working version shows.. 

everything works fine with system set to F, when set to C.. the thermostat shows its 32C+ higher than it should be.. ie 22C = 44C

**Comment by nayrnet** [comment]
sorry bout the close and open, misclicked.. 

I think I may have found the unit inversion:
CPush::ProcessSendValue has a conversion routine in it already:
    if ((vType=="Temperature") || (vType=="Temperature 1") || (vType=="Temperature 2")|| (vType == "Set point"))
    {
        double tvalue=ConvertTemperature(atof(rawsendValue.c_str()),tempsign);
        sprintf(szData,"%.1f", tvalue);
    }

so the mainworker is converting it, then pushing it with this routine thats also converting it.. I think this is converting the conversion and causing the setpoint failure. 

I am adding some debug output to see if I can trace this..

**Comment by nayrnet** [comment]
Current Behavior: both C/F broken
Patch Earlier: F fixed/C broken

from what I can tell ZW devices are getting the input in Celsius, like always.

added some debug output on your patch earlier, right before _tThermostat tmeter; this is what I am seeing:
System set to F, changing setpoint to 70F.. WORKS!
Temp = 70.00

System set to C, changing setpoint to 20C.. FAILS!
Temp = 20.00

System set to C, changing setpoint to 70C.. WORKS but of course SetPoint is actually 70F
Temp = 70.00

So its coming through here unconverted, and skipping the conversion with your patch, then its converted somewhere down the line and sent to the device.. this much I can see.

It appears as if the conversion is indiscriminate, the check to see what unit we are set at is not operating as expected.

Ive searched all around for the math to do the conversion, and I am not finding anything obvious..

**Comment by nayrnet** [comment]
Found that the OpenZwave.cpp does do conversions, see:
https://github.com/domoticz/domoticz/blob/bd5848a5fae3d0157fcfe2f00bb3018b0031767d/hardware/OpenZWave.cpp#L2860

**Comment by nayrnet** [comment]
I give up, I cannot fix this.. using the patch @gizmocuz provided so at least I am working.. sucks.

learned alot, but this is still beyond my abilities.. now to fix my database from all these bs values today.

**Comment by gizmocuz** [comment]
How about:

```
void COpenZWave::SetThermostatSetPoint(const int nodeID, const int instanceID, const int commandClass, const float value)
{
    if (m_pManager == NULL)
        return;
    OpenZWave::ValueID vID(0, 0, OpenZWave::ValueID::ValueGenre_Basic, 0, 0, 0, OpenZWave::ValueID::ValueType_Bool);
    if (GetValueByCommandClass(nodeID, instanceID, COMMAND_CLASS_THERMOSTAT_SETPOINT, vID) == true)
    {
        std::string vUnits = m_pManager->GetValueUnits(vID);
        float temp = value;
        if (vUnits == "F")
        {
            //Convert to Fahrenheit
            temp = (float)ConvertToFahrenheit(temp);
        }
        if (!m_pManager->SetValue(vID, temp))
        {
            _log.Log(LOG_ERROR, "OpenZWave: Error setting Thermostat Setpoint Value! NodeID: %d (0x%02x)", nodeID, nodeID);
        }
    }
    m_updateTime = mytime(NULL);
}
```

will this work ?

**Comment by nayrnet** [comment]
with or without the previous alterations?

with the previous alterations the setpoint is broken nothing works, either C or F.. why converttofahrenheight?

**Comment by gizmocuz** [comment]
without, just as the code is just committed, hopefully this works
Why Fahrenheit, we had to choose a base metric system, and internally everything is in Celsius
so when a zwave node uses F as unit, we convert Celsius to Fahrenheit

**Comment by nayrnet** [comment]
oh so it was just an illusion that my zwave thermostat was operating internally on Celsius when set to Fahrenheit because domoticz is operating internally.. should have figured it out when I discovered the log outputs were converting.. I assumed everything was always Celsius except what was displayed, guess that akes me an ass.. changing my thermostat to match domoticz. (thought I could leave my Thermostat in Fahrenheit with Domo in Celsius)

Seems to have same behavior as before, in F with F thermostat and F Domoticz = Works
in C with C thermostat and C domoticz = Cant change setpoint

If my thermostat and domoticz mismatch units, it dont work.. so when I left my thermostat in Celsius and put domo in Fahrenheit nada, changed thermostat back and it worked.. wonder if i needa restart my controller.

**Comment by nayrnet** [comment]
YAY, it works

Sorry, I guess I did not realize all the steps I should be doing to test Celsius.. Changed Domo to C, changed Thermostat to C, restarted Z-Wave controller, (trigger a bunch of alerts that my house is freezing) let every thing settle down and tested again. 

Worked fine in Celsius too, no longer needed to input a F temp.

man debuging ZW on a big network is a slow going process.. I wonder if I managed to fix it sometime and just never realized it, sigh.
